Added a new Dummy backend for Entity Store

amatriain opened the pull request:
This PR adds the ability to set rack-cache entity store to nil, e.g.:
```
require 'rack/cache'

use Rack::Cache,
  :metastore   => 'file:/var/cache/rack/meta',
  :entitystore => nil,
  :verbose     => true

run app
```
or
```
require 'rack/cache'

use Rack::Cache,
  :metastore   => 'file:/var/cache/rack/meta',
  :entitystore => 'dummy:',
  :verbose     => true

run app
```

Both values (nil or 'dummy:') in the entitystore option enable the Dummy backend.

When the Dummy backend is used for the entity store, rack-cache doesn't persist response bodies. If a cached response is still fresh, or if it's stale but found to be still valid, rack-cache returns only the response headers with an empty body. The code performing the request should check if the response comes from the cache (e.g. checking the X-Rack-Cache header) and in in this case ignore the response body.

Obviously this is only useful for use cases in which, if a cached response is still valid, its body is not interesting.

This is useful for my particular use case but may be interesting for other people. In particular this avoids the common problem of the backend (disk in my case) filling up with old entities, because rack-cache performs no cleanup of old useless entities once there is a new response. This can be avoided with this PR by not persisting entries in the backend in the first place.

grosser commented on lib/rack/cache/entitystore.rb:
```diff
@@ -336,6 +336,38 @@ def self.resolve(uri)
+      end
+
+      def open(key)
+        nil
```
can this return `[]` to get rid of the conditions in metastore.rb ?

amatriain replied:
Not without changing ```MetaStore#store```. 

For some reason I don't understand, line 69 of that method does this:

```
response.body = entity_store.open(digest) || response.body
```

This is, just after writing the response in the cache, the body is read from the cache and the response body is overwritten with whatever is read from the entity store, unless ```entity_store.open(digest)``` returns nil. 

If ```Dummy#open``` returns ```[ ]``` (or anything other than nil, really), the response body will be overwritten with ```[ ]``` even for requests that miss the cache. This means that the actual response body will **never** be returned by the library, not even for the first request (which results in a cache miss and the response being stored in the cache).

The only solutions are either to make ```Dummy.open``` return nil so that the response body is not overwritten, or to get rid of line 69. If we remove line 69 of ```MetaStore```, ```Dummy``` can return ```[ ]``` instead of nil.

I can't see the reason for that line, I don't think it makes sense to read the response body from the cache immediately after saving it in the cache. But I may be missing something important there. What do you think?

amatriain replied:
I've finally decided that you're right and it's better to return ```''``` from ```Noop#read``` and ```[]``` from ```Noop#open```. This brings it in line with other backends, which always return objects suitable as Rack bodies from ```#open``` (an empty array can be a Rack body, ```nil``` can't because it doesn't respond to ```each```).

This unfortunately requires a change to ```MetaStore#store``` so that the line
```
response.body = entity_store.open(digest) || response.body
```
is not executed if the entitystore backend is a Noop.

The commit with this change is forthcoming.

amatriain commented:
@grosser you may like this PR better than the previous one. It mostly implements the feature by adding a new backend class to Entity Store.

A couple changes to the core code have still been unavoidable. They should be very minor.

grosser commented on lib/rack/cache/entitystore.rb:
```diff
@@ -336,6 +336,38 @@ def self.resolve(uri)
+    end
+
+    NIL = Dummy
+    DUMMY = Dummy
```
why do we need 2 constants ?

amatriain replied:
To allow using this backend either with ```entitystore: nil``` or ```entitystore: 'dummy:'```

Heap, Mem, Disk, File, Gaecache and Gae are the same. I followed the same style.

grosser commented on lib/rack/cache/storage.rb:
```diff
@@ -38,6 +38,9 @@ def create_store(type, uri)
         end
+      elsif uri.nil?
+        # hack to support setting entitystore to nil, which doesn't persist request bodies
+        type.const_get('NIL').resolve(uri)
```
call Dummy.new here ?
maybe set the uri to Dummy.new earlier / when initializing ?

amatriain replied:
All other subclasses of EntityStore are instanced like this, calling ```Foo.resolve(uri)``` which returns the new instance. There is no call to ```Foo.new``` to instance an entity store in ```Storage.create_store```. I just followed the same style.

If what you dislike is the use if the magic string ```'NIL'``` I can change it to this: 

```
elsif uri.nil?
        # hack to support setting entitystore to nil, which doesn't persist request bodies
        type.const_get(uri.inspect.upcase).resolve(uri)
```

Better?

grosser replied:
const_get + user-supplied.inspect smells like trouble ...  know it's safe,
but just looks weird :D

On Tue, Jan 26, 2016 at 4:19 PM, Alfredo Amatriain <notifications@github.com
> wrote:

> In lib/rack/cache/storage.rb
> <https://github.com/rtomayko/rack-cache/pull/130#discussion_r50924182>:
>
> > @@ -38,6 +38,9 @@ def create_store(type, uri)
> >          else
> >            fail "Unknown storage provider: #{uri.to_s}"
> >          end
> > +      elsif uri.nil?
> > +        # hack to support setting entitystore to nil, which doesn't persist request bodies
> > +        type.const_get('NIL').resolve(uri)
>
> All other subclasses of EntityStore are instanced like this, calling
> Foo.resolve(uri) which returns the new instance. There is no call to
> Foo.new to instance an entity store in Storage.create_store. I just
> followed the same style.
>
> If what you dislike is the use if the magic string 'NIL' I can change it
> to this:
>
> elsif uri.nil?
>         # hack to support setting entitystore to nil, which doesn't persist request bodies
>         type.const_get(uri.inspect.upcase).resolve(uri)
>
> Better?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rtomayko/rack-cache/pull/130/files#r50924182>.
>

grosser replied:
l'll take a closer look later ... maybe this is as good as it gets ...

On Tue, Jan 26, 2016 at 4:25 PM, Michael Grosser <michael@grosser.it> wrote:

> const_get + user-supplied.inspect smells like trouble ...  know it's safe,
> but just looks weird :D
>
> On Tue, Jan 26, 2016 at 4:19 PM, Alfredo Amatriain <
> notifications@github.com> wrote:
>
>> In lib/rack/cache/storage.rb
>> <https://github.com/rtomayko/rack-cache/pull/130#discussion_r50924182>:
>>
>> > @@ -38,6 +38,9 @@ def create_store(type, uri)
>> >          else
>> >            fail "Unknown storage provider: #{uri.to_s}"
>> >          end
>> > +      elsif uri.nil?
>> > +        # hack to support setting entitystore to nil, which doesn't persist request bodies
>> > +        type.const_get('NIL').resolve(uri)
>>
>> All other subclasses of EntityStore are instanced like this, calling
>> Foo.resolve(uri) which returns the new instance. There is no call to
>> Foo.new to instance an entity store in Storage.create_store. I just
>> followed the same style.
>>
>> If what you dislike is the use if the magic string 'NIL' I can change it
>> to this:
>>
>> elsif uri.nil?
>>         # hack to support setting entitystore to nil, which doesn't persist request bodies
>>         type.const_get(uri.inspect.upcase).resolve(uri)
>>
>> Better?
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/rtomayko/rack-cache/pull/130/files#r50924182>.
>>
>
>

grosser commented on lib/rack/cache/entitystore.rb:
```diff
@@ -336,6 +336,38 @@ def self.resolve(uri)
     GAECACHE = GAEStore
     GAE = GAEStore
 
+    # Dummy Entity Store that does not persist entities
```
some usage instructions would be nice

grosser commented:
looks like a lot less conditions then the previous one :)

amatriain commented:
Added better comments for the Dummy backend. Also removed magic string "NIL" from the instanciation of the Dummy backend, use ```uri.inspect.upcase``` instead.

grosser commented on lib/rack/cache/storage.rb:
```diff
@@ -38,6 +38,9 @@ def create_store(type, uri)
+        type.const_get(uri.inspect.upcase).resolve(uri)
       else
         # hack in support for passing a Dalli::Client or Memcached object
         # as the storage URI.
```
hmm how about we stop doing this complex string mapping and accept any instance,
so this can be `entitystore: Rack::Cache::EntityStore::Noop.new` and we don't need any parsing ?

grosser replied:
or Nil ... or Nothing ... Dummy sounds a bit vague to me ...

amatriain replied:
The way backends are instanced in ```Storage#create_store``` is very convoluted, in my opinion. Changing it may make sense, but I think it's outside the scope of this PR. Maybe in another PR.

Changing the way options are passed from clients, so that class instances are accepted instead of uri strings, is in my opinion a bad idea. First, it would break any code that uses this gem if backwards compatibility is dropped; if it is maintained, backend instancing would have to be twice as complicated as it is now. Second, it would expose too much of the gem's internals to client code: why should clients need to be aware that the class that implements file backend is ```Rack::Cache::EntityStore::Disk```?

amatriain replied:
```Dummy``` doesn't sound right to me either. ```Noop``` sounds better, I like it better than ```Nil``` or ```Nothing```.

grosser commented:
this is looking really good :heart: how about some tests ? :)

grosser commented on lib/rack/cache/metastore.rb:
```diff
@@ -64,7 +64,8 @@ def store(request, response, entity_store)
         response.headers['Content-Length'] = size.to_s unless response.headers['Transfer-Encoding']
-        response.body = entity_store.open(digest) || response.body
+        # If the entitystore backend is a Noop, do not try to read the body from the backend, it always returns an empty array
+        response.body = entity_store.open(digest) || response.body unless entity_store.is_a? Rack::Cache::EntityStore::Noop
```
why do we need the condition, open returns `[]` ?

grosser commented on lib/rack/cache/storage.rb:
```diff
@@ -38,6 +38,9 @@ def create_store(type, uri)
         else
           fail "Unknown storage provider: #{uri.to_s}"
         end
+      elsif uri.nil?
```
this is no longer needed since `noop://` will go through the first case ?

grosser commented on lib/rack/cache/entitystore.rb:
```diff
@@ -336,6 +336,45 @@ def self.resolve(uri)
+      end
+    end
+
+    NIL = Noop
```
do we still need these constants when using `noop://` ?
